drivers/flash/nrf_qspi_nor.c

```diff
@@ -0,0 +1,687 @@
+{
+	struct qspi_nor_data *dev_data = get_dev_data(dev);
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+}
+
+static inline void qspi_unlock(struct device *dev)
+{
+	struct qspi_nor_data *dev_data = get_dev_data(dev);
+
+	k_sem_give(&dev_data->sem);
+}
+
+static inline void qspi_wait_for_completion(struct device *dev)
+{
+	struct qspi_nor_data *dev_data = get_dev_data(dev);
+
+	k_sem_take(&dev_data->sync, K_FOREVER);
+}
+
+static inline void qspi_complete(struct device *dev)
+{
+	struct qspi_nor_data *dev_data = get_dev_data(dev);
+
+	k_sem_give(&dev_data->sync);
+}
+
+/**
+ * @brief QSPI handler
+ *
+ * @param event Driver event type
+ * @param p_context Pointer to context. Use in interrupt handler.
+ * @retval None
+ */
+static void qspi_handler(nrfx_qspi_evt_t event, void *p_context)
+{
+	struct device *dev = p_context;
+
+	if (event == NRFX_QSPI_EVENT_DONE) {
+		qspi_complete(dev);
+		qspi_unlock(dev);
+	}
+}
+
+
+/* QSPI send custom command */
+static int qspi_send_cmd(struct device *dev, const struct qspi_cmd *cmd)
+{
+	/* Check input parameters */
+	if (!cmd) {
+		return -EINVAL;
+	}
+
+	qspi_lock(dev);
+
+	nrf_qspi_cinstr_conf_t cinstr_cfg = {
+		.opcode = cmd->op_code,
+		.io2_level = true,
+		.io3_level = true,
+		.wipwait = false,
+		.wren = true,
+	};
+	cinstr_cfg.length = sizeof(cmd->op_code);
+	if ((cmd->tx_buf != 0) && (cmd->rx_buf != 0)) {
+		cinstr_cfg.length += cmd->tx_buf->len + cmd->rx_buf->len;
+	} else if ((cmd->tx_buf != 0) && (cmd->rx_buf == 0)) {
+		cinstr_cfg.length += cmd->tx_buf->len;
+	} else if ((cmd->tx_buf == 0) && (cmd->rx_buf != 0)) {
+		cinstr_cfg.length += cmd->rx_buf->len;
+	}
+
+	int res = nrfx_qspi_cinstr_xfer(&cinstr_cfg, cmd->tx_buf->buf,
+					cmd->rx_buf->buf);
+
+	qspi_unlock(dev);
+	return qspi_get_zephyr_ret_code(res);
+}
+
+/* QSPI erase */
+static int qspi_erase(struct device *dev, u32_t addr, u32_t size)
+{
+	/* Check input parameters */
+	if (!size) {
+		return -EINVAL;
+	}
+
+	int res = 0;
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	if ((addr + size) <= params->size) {
+		while (size) {
+			if (size == params->size) {
+				/* chip erase */
+				qspi_lock(dev);
+				res = nrfx_qspi_chip_erase();
+				qspi_wait_for_completion(dev);
+				size = 0;
+			} else if ((size >= QSPI_BLOCK_SIZE) &&
+				   QSPI_IS_BLOCK_ALIGNED(addr)) {
+				/* 64 kB block erase */
+				qspi_lock(dev);
+				res = nrfx_qspi_erase(NRF_QSPI_ERASE_LEN_64KB, addr);
+				qspi_wait_for_completion(dev);
+				addr += QSPI_BLOCK_SIZE;
+				size -= QSPI_BLOCK_SIZE;
+			} else if ((size >= QSPI_SECTOR_SIZE) &&
+				   QSPI_IS_SECTOR_ALIGNED(addr)) {
+				/* 4kB sector erase */
+				qspi_lock(dev);
+				res = nrfx_qspi_erase(NRF_QSPI_ERASE_LEN_4KB, addr);
+				qspi_wait_for_completion(dev);
+				addr += QSPI_SECTOR_SIZE;
+				size -= QSPI_SECTOR_SIZE;
+			} else {
+				/* minimal erase size is at least a sector size */
+				LOG_DBG("unsupported at 0x%lx size %zu", (long)addr, size);
```

Comment:
Errors should be rather reported with `LOG_ERR`.

Reply:
Changed `LOG_DBG` to `LOG_ERR`.